### PR TITLE
Saving contract data in localStorage.

### PR DESCRIPTION
**Pull request by merivale**
This PR extends the dummy Marlowe capability to use localStorage to save some data that we will (when it's working) get from the PAB. This makes it possible to keep the rest of the code closer to what it will be wgen the PAB is integrated, and - in the meantime - let's us run simulations with roles and multiple wallets (the localStorage is used to share the contract data between different wallets).

My dream is to get this working so that you can have different wallets open in different tabs, and have them all just automatically stay in sync. I tried polling the localStorage every second, and that made the interface noticeably sluggish. So I think I'll need to listen for localStorage events and only update when an item is set. Going to try that tomorrow...

### CODE REVIEW

File: marlowe-dashboard-client/src/Play/State.purs [hunk 1]
```diff
@@ -173,6 +175,43 @@ handleAction _ CloseCard = do
   for_ (init cards) \remainingCards ->
     assign _cards remainingCards
 
+-- FIXME: until the PAB is working, we get the latest data from local storage here (this action
+-- is called every second)
+handleAction { currentSlot } UpdateFromStorage = do
```
Comment (hrajchert) [inline marlowe-dashboard-client/src/Play/State.purs:180]: I think it's better to add a subscription to local storage events rather than polling

**Review by hrajchert — COMMENTED** [review]
Looking good, minor details

File: marlowe-dashboard-client/src/Capability/LocalStorage.purs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+instance manageLocalStorageAppM :: ManageLocalStorage AppM where
+  getWalletLibrary = do
+    mWalletLibraryJson <- liftEffect $ getItem walletLibraryLocalStorageKey
+    pure
+      $ case mWalletLibraryJson of
+          Just json -> case runExcept $ decodeJSON json of
+            Right wallets -> wallets
+            Left _ -> mempty
+          Nothing -> mempty
```
[the remark this answers is not in the document]
Comment (merivale) [reply marlowe-dashboard-client/src/Capability/LocalStorage.purs:54]: ah cool, I didn't know about `hush`

File: marlowe-dashboard-client/src/Marlowe/Execution.purs [hunk 3]
```diff
@@ -140,7 +140,7 @@ timeoutState currentSlot { current, previous, mPendingTimeouts, mNextTimeout } =
       Contract ->
       { mNextTimeout :: Maybe Slot, mPendingTimeouts :: Maybe PendingTimeouts }
     advanceAllTimeouts mNextTimeout' newTimeouts state' contract'
-      | mNextTimeout' >= Just currentSlot =
+      | mNextTimeout' /= Nothing && mNextTimeout' < Just currentSlot =
```
[the remark this answers is not in the document]
Comment (merivale) [reply marlowe-dashboard-client/src/Marlowe/Execution.purs:143]: :thinking:

File: marlowe-dashboard-client/src/Play/State.purs (hunk 1 above)
Comment (merivale) [reply marlowe-dashboard-client/src/Play/State.purs:180]: yep, that's my plan for today - this is an outdated comment from when I was exprimenting with polling (it didn't work well :upside_down_face:)

**Pull request merged**